File: src/64_Accumulate/64_Accumulate.py
class Solver:
    """解法类的基类"""

    def __init__(self):
        self.result = 0

    def sum(self, n: int) -> int:
        pass

    def reset(self) -> None:
        self.result = 0


# ==============================
# 解法1：利用逻辑运算的短路效应+递归
class Solver1(Solver):
    def sum(self, n: int) -> int:
        (not not n) and self.sum(n - 1)
        self.result += n
        return self.result


# =============================
# 解法2：利用类数组实现递归（对应于C++中的虚函数）


class Solver2_A:
    """基类"""

    def getSum(self, n: int) -> int:
        return 0


class Solver2_B(Solver2_A):
    """子类"""

    def __init__(self, arr: list[Solver2_A]):
        self.arr: list[Solver2_A] = arr

    def getSum(self, n: int) -> int:
        # 利用短路运算充当真假判断：!n==false, !!n==true
        return n + self.arr[not not n].getSum(n - 1)


class Solver2(Solver):
    def __init__(self):
        super().__init__()

        # 定义存储两种情况的数组
        self.cls_l = [Solver2_A(), Solver2_B([])]
        self.cls_l[1].arr = self.cls_l

    def sum(self, n: int) -> int:
        return self.cls_l[1].getSum(n)  # 递归的起点


# ===============================
# 解法3：利用函数数组实现递归（对应于C++中的函数指针）


class Solver3(Solver):
    def __init__(self):
        super().__init__()
        # 利用python的函数闭包轻松构建函数数组
        self.f = [self.terminate, self.recur]

    def sum(self, n: int) -> int:
        return self.recur(n)

    def recur(self, n: int) -> int:
        return n + self.f[not not n](n - 1)

    def terminate(self, n: int) -> int:
        return 0


# ===============================
class Solution:
    def __init__(self) -> None:
        self.solvers: list[Solver] = [
            None,
            Solver1(),
            Solver2(),
            Solver3(),
        ]

    # solver_i取值范围为1到3，对应3种解法
    def sumNums(self, n: int, solver_i: int = 3) -> int:
        if solver_i:
            solver = self.solvers[solver_i]
            solver.reset()
            return solver.sum(n)
        else:
            return 0


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, n: int, expected: int) -> None:
        text = testName
        result = self.solution.sumNums(n)

        text += f": result={result}, expected={expected},"

        if result == expected:
            text += " passed."
        else:
            text += " failed."
        print(text)

    def runTest(self) -> None:
        self.checkResult("Test1", 1, 1)
        self.checkResult("Test2", 5, 15)
        self.checkResult("Test3", 10, 55)
        self.checkResult("Test4", 0, 0)


if __name__ == "__main__":
    test = Test()
    test.runTest()

File: src/64_Accumulate/64_Accumulate.cpp
// 面试题64：求1+2+…+n
// 题目：求1+2+…+n，要求不能使用乘除法、for、while、if、else、switch、case
// 等关键字及条件判断语句（A?B:C）。

#include <iostream>
#include <string>
#include <vector>

using namespace std;

namespace Solve {

class Solver {
   public:
    virtual int sum( int n ) = 0;
};

// ===============================
// 解法1：利用逻辑运算的短路效应+递归

class Solver1 : public Solver {
   public:
    int sum( int n ) override {
        // 利用短路运算充当真假判断：!n==false, !!n==true
        !!n && ( n += sum( n - 1 ) );
        return n;
    }
};

// =============================
// 解法2：利用构造函数+静态变量
class Solver2Temp {
   public:
    Solver2Temp() { N++, Sum += N; }

    static void reset() { N = 0, Sum = 0; }
    static int getSum() { return Sum; }

   private:
    static int N, Sum;  // 定义静态变量
};

int Solver2Temp::N = 0, Solver2Temp::Sum = 0;  // 静态变量需要初始化

class Solver2 : public Solver {
   public:
    int sum( int n ) override {
        Solver2Temp::reset();
        // 利用构造数组来替代循环
        Solver2Temp *tmp = new Solver2Temp[ n ];
        delete[] tmp;
        tmp = nullptr;
        return Solver2Temp::getSum();
    }
};

// ===========================
// 解法3：利用虚函数替代递归
class Solver3_A {
   public:
    virtual int getSum( int n ) { return 0; }
};

Solver3_A *solver3Arr[ 2 ];  // 替代递归的关键，在于定义存储两种情况的数组

class Solver3_B : public Solver3_A {
   public:
    virtual int getSum( int n ) {
        // 利用短路运算充当真假判断：!n==false, !!n==true
        // 根据真假在父类或子类的getSum函数中选择，从而实现递归
        return solver3Arr[ !!n ]->getSum( n - 1 ) + n;
    }
};

class Solver3 : public Solver {
   public:
    int sum( int n ) override {
        Solver3_A a;  // 父类
        Solver3_B b;  // 子类
        solver3Arr[ 0 ] = &a, solver3Arr[ 1 ] = &b;
        return solver3Arr[ 1 ]->getSum( n );  // 递归的起点
    }
};

// ====================================
// 解法4：利用函数指针，替代虚函数

typedef int ( *Solver4Func )( int );

int solver4Terminate( int n ) { return 0; }

int solver4GetSum( int n ) {
    // 与虚函数法的思路类似，用一个数组存储两类情况下的递归函数
    // 类内的函数指针（如以下列表内的）必须调用，因此此函数必须在类外定义
    static Solver4Func f[ 2 ] = { solver4Terminate, solver4GetSum };
    // 利用短路运算充当真假判断：!n==false, !!n==true
    return n + f[ !!n ]( n - 1 );
}

class Solver4 : public Solver {
   public:
    int sum( int n ) override { return solver4GetSum( n ); }
};

}  // namespace Solve


// ===============================
class Solution {
   public:
    // solver_i取值范围为1到4，对应4种解法
    int sumNums( int n, int solver_i = 1 ) {
        if ( solver_i )
            return solvers[ solver_i ]->sum( n );
        else
            return 0;
    }

    Solution() {
        solvers = SolVec{
            nullptr,
            new Solve::Solver1(),
            new Solve::Solver2(),
            new Solve::Solver3(),
            new Solve::Solver4(),
        };
    }

   private:
    using Solver = Solve::Solver;
    using SolVec = vector<Solver *>;

    SolVec solvers;
};

class Test {
   public:
    void checkResult( const string testName, int n, int expected ) {
        cout << testName;
        int result = solution.sumNums( n );

        if ( result == expected )
            cout << " passed.\n";
        else
            cout << " failed.\n";
    }

    void runTest() {
        checkResult( "Test1", 1, 1 );
        checkResult( "Test2", 5, 15 );
        checkResult( "Test3", 10, 55 );
        checkResult( "Test4", 0, 0 );
    }

   private:
    Solution solution;
};

int main() {
    Test test;
    test.runTest();
    return 0;
}